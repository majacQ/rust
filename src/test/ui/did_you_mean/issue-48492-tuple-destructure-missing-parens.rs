// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

#![allow(unused)]

#[derive(Copy, Clone)]
enum Nucleotide {
    Adenine,
    Thymine,
    Cytosine,
    Guanine
}

#[derive(Clone)]
struct Autosome;

#[derive(Clone)]
enum Allosome {
    X(Vec<Nucleotide>),
    Y(Vec<Nucleotide>)
}

impl Allosome {
    fn is_x(&self) -> bool {
        match *self {
            Allosome::X(_) => true,
            Allosome::Y(_) => false,
        }
    }
}

#[derive(Clone)]
struct Genome {
    autosomes: [Autosome; 22],
    allosomes: (Allosome, Allosome)
}

fn find_start_codon(strand: &[Nucleotide]) -> Option<usize> {
    let mut reading_frame = strand.windows(3);
    while let b1, b2, b3 = reading_frame.next().expect("there should be a start codon") {
        //~^ ERROR unexpected `,` in pattern
        // ...
    }
    None
}

fn find_thr(strand: &[Nucleotide]) -> Option<usize> {
    let mut reading_frame = strand.windows(3);
    let mut i = 0;
    if let b1, b2, b3 = reading_frame.next().unwrap() {
        //~^ ERROR unexpected `,` in pattern
        match (b1, b2, b3) {
            //~^ ERROR cannot find value `b2` in this scope
            //~| ERROR cannot find value `b3` in this scope
            Nucleotide::Adenine, Nucleotide::Cytosine, _ => {
                //~^ ERROR unexpected `,` in pattern
                return Some(i);
            },
            _ => {}
        }
        i += 1;
    }
    None
}

fn analyze_sex_chromosomes(women: &[Genome], men: &[Genome]) {
    for x, _barr_body in women.iter().map(|woman| woman.allosomes.clone()) {
        //~^ ERROR unexpected `,` in pattern
        // ...
    }
    for x, y @ Allosome::Y(_) in men.iter().map(|man| man.allosomes.clone()) {
        //~^ ERROR unexpected `,` in pattern
        // ...
    }
}

fn main() {
    let genomes = Vec::new();
    let women, men: (Vec<Genome>, Vec<Genome>) = genomes.iter().cloned()
    //~^ ERROR unexpected `,` in pattern
        .partition(|g: &Genome| g.allosomes.0.is_x() && g.allosomes.1.is_x());
}
