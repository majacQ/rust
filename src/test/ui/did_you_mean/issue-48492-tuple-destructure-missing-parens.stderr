error: unexpected `,` in pattern
  --> $DIR/issue-48492-tuple-destructure-missing-parens.rs:47:17
   |
47 |     while let b1, b2, b3 = reading_frame.next().expect("there should be a start codon") {
   |               --^------- help: try adding parentheses: `(b1, b2, b3)`

error: unexpected `,` in pattern
  --> $DIR/issue-48492-tuple-destructure-missing-parens.rs:57:14
   |
57 |     if let b1, b2, b3 = reading_frame.next().unwrap() {
   |            --^------- help: try adding parentheses: `(b1, b2, b3)`

error: unexpected `,` in pattern
  --> $DIR/issue-48492-tuple-destructure-missing-parens.rs:62:32
   |
62 |             Nucleotide::Adenine, Nucleotide::Cytosine, _ => {
   |             -------------------^------------------------ help: try adding parentheses: `(Nucleotide::Adenine, Nucleotide::Cytosine, _)`

error: unexpected `,` in pattern
  --> $DIR/issue-48492-tuple-destructure-missing-parens.rs:74:10
   |
74 |     for x, _barr_body in women.iter().map(|woman| woman.allosomes.clone()) {
   |         -^----------- help: try adding parentheses: `(x, _barr_body)`

error: unexpected `,` in pattern
  --> $DIR/issue-48492-tuple-destructure-missing-parens.rs:78:10
   |
78 |     for x, y @ Allosome::Y(_) in men.iter().map(|man| man.allosomes.clone()) {
   |         -^------------------- help: try adding parentheses: `(x, y @ Allosome::Y(_))`

error: unexpected `,` in pattern
  --> $DIR/issue-48492-tuple-destructure-missing-parens.rs:86:14
   |
86 |     let women, men: (Vec<Genome>, Vec<Genome>) = genomes.iter().cloned()
   |         -----^---- help: try adding parentheses: `(women, men)`

error[E0425]: cannot find value `b2` in this scope
  --> $DIR/issue-48492-tuple-destructure-missing-parens.rs:59:20
   |
59 |         match (b1, b2, b3) {
   |                    ^^ did you mean `b1`?

error[E0425]: cannot find value `b3` in this scope
  --> $DIR/issue-48492-tuple-destructure-missing-parens.rs:59:24
   |
59 |         match (b1, b2, b3) {
   |                        ^^ did you mean `b1`?

error: aborting due to 8 previous errors

